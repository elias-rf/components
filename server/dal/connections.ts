import { TConnections } from "../../types";
import fullvision from "./fullvision.connection";
import oftalmo from "./oftalmo.connection";
import plano from "./plano.connection";

export const connections = {
  oftalmo,
  plano,
  fullvision,
};

// export type TConnections = typeof connections;
export type TDbs = keyof TConnections;
