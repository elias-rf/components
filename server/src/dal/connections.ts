import oftalmo from "./oftalmo.connection";
import plano from "./plano.connection";
import fullvision from "./fullvision.connection";

const connections = {
  oftalmo,
  plano,
  fullvision,
};

export type Connections = typeof connections;

export default connections;
