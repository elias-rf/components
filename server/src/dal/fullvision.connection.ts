import KnexObj, { Knex } from "knex";
import knexLogger from "../lib/knex/knex-logger";
import { optionsFullvision } from "../config/knex";
import config from "../config";

let knexFullvision: Knex;

if (config.app.env === "test") {
  knexFullvision = KnexObj(optionsFullvision);
} else {
  knexFullvision = knexLogger(KnexObj(optionsFullvision));
}

export default knexFullvision;
