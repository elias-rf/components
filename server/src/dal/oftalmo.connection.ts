import KnexObj, { Knex } from "knex";
import config from "../config";
import { optionsOftalmo } from "../config/knex";
import knexLogger from "../lib/knex/knex-logger";

let knexOftalmo: Knex;

if (config.app.env === "test") {
  knexOftalmo = KnexObj(optionsOftalmo);
} else {
  knexOftalmo = knexLogger(KnexObj(optionsOftalmo));
}

export default knexOftalmo;
