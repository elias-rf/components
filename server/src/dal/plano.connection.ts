import KnexObj, { Knex } from "knex";
import knexLogger from "../lib/knex/knex-logger";
import { optionsPlano } from "../config/knex";
import config from "../config";

let knexPlano: Knex;

if (config.app.env === "test") {
  knexPlano = KnexObj(optionsPlano);
} else {
  knexPlano = knexLogger(KnexObj(optionsPlano));
}
export default knexPlano;
