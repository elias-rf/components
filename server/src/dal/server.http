### API root
GET http://localhost:3000/api

### Login
POST http://localhost:3000/api/login
Content-Type: application/json

{"user":"elias","password":"jkl"}

### WhoAmI
GET http://localhost:3000/api/whoami

### Logout
POST http://localhost:3000/api/logout

### Schemas do Local
OPTIONS http://localhost:3000/api/local

### Schemas da tabela
OPTIONS http://localhost:3000/api/oftalmo/phonebook

### Schemas da tabela
GET http://localhost:3000/api/oftalmo/phonebook


### Autorizações com chave primária composta
GET http://localhost:3000/api/local/authorization/1/usuarios

### Authorization
GET http://localhost:3000/api/local/authorization

### Colunas da tabela
GET http://localhost:3000/api/oftalmo/phonebook

### Colunas da tabela
GET http://localhost:3000/api/oftalmo/phonebook/count

### api list
GET http://localhost:3000/api/oftalmo/phonebook
  ?order={"id":"asc"}
  &select=id,name
  &where={"id =":"105"}
  &limit=2

### api read
GET http://localhost:3000/api/oftalmo/phonebook/99

### api create
POST http://localhost:3000/api/oftalmo/phonebook
Content-Type: application/json

{"id":99,"name":"fulano"}

### api update
PUT http://localhost:3000/api/oftalmo/phonebook/99
Content-Type: application/json

{ "name":"fulano2"}

### api delete
DELETE http://localhost:3000/api/oftalmo/phonebook/99

### api list
GET http://localhost:3000/api/local/subject

### api list
GET http://localhost:3000/api/oftalmo/user

### api list
GET http://localhost:3000/api/local/authorization?order={"idUser":"asc"}

### api list
GET http://localhost:3000/api/oftalmo/operacaoMensal?inicio=2021-01&operacao=2018

### api list
GET http://localhost:3000/api/oftalmo/operacaoDiario?inicio=2020-01-01&fim=2020-01-31&operacao=2018

### api list
GET http://localhost:3000/api/oftalmo/operacaoProduto?inicio=2021-02-02&operacao=2018

### api list
GET http://localhost:3000/api/oftalmo/operacaoTurno?inicio=2021-02-02&operacao=2018

### api list
GET http://localhost:3000/api/oftalmo/operacaoModelo?inicio=2021-02-02&operacao=2018&produto=liteflex

### api list
GET http://localhost:3000/api/plano/transferenciaMensal?inicio=2020-10

### api list
GET http://localhost:3000/api/plano/transferenciaDiario?inicio=2020-01-01&fim=2020-01-31

### api list
GET http://localhost:3000/api/plano/transferenciaModelo?inicio=2021-02-02

###
GET http://localhost:3000/api/oftalmo/operacaoMensal?mes=2020-02&operacao=1010

###
GET http://localhost:3000/api/oftalmo/user/1?where={"ativo =":"1"}

###
POST http://localhost:3000/api/qs
Content-Type: application/json

{
  "where": {
    "id": {"<=":99, ">=":1},
    "name": {"=":"fulano"}
  }
}

###
GET http://localhost:3000/api/qs
  ?q={"select":["f1","f2"],"order":[{"column":"id","order":"asc"}],"where":[["id","<=",99],["id",">=",1]]}
