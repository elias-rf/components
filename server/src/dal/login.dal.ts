import config from "../config/index";
import jwtEncode from "../lib/jwt-encode";
import passwordVerify from "../lib/password-verify";
import { Connections } from "./connections";

async function login(
  { user, password }: { user: string; password: string },
  { oftalmo }: Connections
) {
  let record: {
    kUsuario: string;
    NomeUsuario: string;
    Ativo: number;
    hash: string;
    nome: string;
    idGroup: string;
  };
  const knex = oftalmo;

  if (!user) {
    throw new Error("Usuário não informado");
  }

  if (!password) {
    throw new Error("Senha não informada");
  }

  try {
    const rec = await knex("tbl_seguranca_usuario").where("NomeUsuario", user);
    record = rec[0];
  } catch (err: any) {
    throw new Error(`erro de acesso ao banco de dados: ${err.message}`);
  }

  if (!record) {
    throw new Error("Usuário não cadastrado");
  }

  if (record.Ativo === 0) {
    throw new Error("Cadastro inativo");
  }

  const passOk = passwordVerify(password, record.hash, "sha256");
  if (!passOk) {
    throw new Error("Senha incorreta");
  }

  const resp: {
    kUsuario: string;
    NomeUsuario: string;
    nome: string;
    idGroup: string;
    token?: string;
  } = {
    kUsuario: record.kUsuario,
    NomeUsuario: record.NomeUsuario,
    nome: record.nome,
    idGroup: record.idGroup,
  };
  resp.token = jwtEncode(resp, config.auth.secret, config.auth.expiration);
  return resp;
}

export default login;
