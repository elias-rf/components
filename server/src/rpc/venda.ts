import { Connections } from "dal/connections";
import createError from "http-errors";
import { CurrentUser } from "../..";
import { Id } from "../../..";

export default function ClienteRpc(connection: Connections) {
  const knexPlano = connection.plano;
  return {
    async diario(
      { inicio, fim, uf }: { inicio: string; fim: string; uf: string[] },
      ctx?: { currentUser: CurrentUser }
    ) {
      if (!inicio) throw new createError.BadRequest("Início não informado");
      if (!fim) throw new createError.BadRequest("Fim não informado");
      if (!uf) throw new createError.BadRequest("UF não informado");

      const qry = knexPlano(
        knexPlano.raw(
          "MestreNota, ItemNota, CadVen, NatOpe, CadPro, CadCli, CategPro, CidadesERF"
        )
      )
        .select([
          "NmCategoria",
          "MestreNota.DtEmissao",
          knexPlano.raw(
            "sum(ItemNota.Quantidade) as quantidade, sum(itemNota.Quantidade * (case MestreNota.tipo when 'E' then case mestrenota.cdcliente when 3158 then ItemNota.VlLiquido * -3 else ItemNota.VlLiquido * -1 end when 'S' then case mestrenota.cdcliente when 3158 then ItemNota.VlLiquido * 3 else ItemNota.VlLiquido end end)) as valor"
          ),
        ])
        .where({
          "MestreNota.FgEstatistica": "S",
          "MestreNota.CdFilial": 2,
          "CadPro.FgEstatistica": "S",
          "CadVen.FgControle": "S",
          "ItemNota.ImprimeComponentes": "N",
        })
        .where("ItemNota.Sequencia", ">", 0)
        .where("MestreNota.Tipo", "<>", "C")
        .whereBetween("MestreNota.DtEmissao", [inicio, fim])
        .whereRaw(
          "ItemNota.CdFilial = MestreNota.CdFilial and ItemNota.Serie = MestreNota.Serie and ItemNota.Modelo = MestreNota.Modelo and ItemNota.NumNota = MestreNota.NumNota and CadPro.CdCategoria = CategPro.CdCategoria and NatOpe.Nop = MestreNota.Nop and CadVen.CdVendedor = MestreNota.CdVendedor and CadCli.CdCliente = MestreNota.CdCliente and CadPro.CdProduto = ItemNota.CdProduto and Cadcli.Cidade = cidadeserf.NmCidadeIBGE"
        )
        .whereIn("cidadeserf.uf", uf)
        .groupBy("NmCategoria", "MestreNota.DtEmissao")
        .orderBy("NmCategoria")
        .orderBy("MestreNota.DtEmissao", "desc");
      return qry;
    },

    async mensal(
      { inicio, fim, id }: { inicio: string; fim: string; id: Id },
      ctx?: { currentUser: CurrentUser }
    ) {
      if (!inicio) throw new createError.BadRequest("Início não informado");
      if (!fim) throw new createError.BadRequest("Fim não informado");
      if (!id) throw new createError.BadRequest("ID não informado");

      const qry = knexPlano(
        knexPlano.raw("MestreNota, ItemNota, CadVen, NatOpe, CadPro, CategPro")
      )
        .select([
          "CategPro.NmCategoria",
          "MestreNota.CdCliente",
          knexPlano.raw(
            "CONVERT(char(7), MestreNota.DtEmissao, 126) AS anoMes, sum(case MestreNota.tipo when 'E' then ItemNota.Quantidade * -1 when 'S' then ItemNota.Quantidade end) as quantidade, sum(itemNota.VlLiquido * (case MestreNota.tipo when 'E' then ItemNota.Quantidade * -1 when 'S' then ItemNota.Quantidade end)) AS valor"
          ),
        ])

        .where({
          "MestreNota.FgEstatistica": "S",
          "MestreNota.CdFilial": 2,
          "CadPro.FgEstatistica": "S",
          "CadVen.FgControle": "S",
          "ItemNota.ImprimeComponentes": "N",
        })
        .where("ItemNota.Sequencia", ">", 0)
        .where("MestreNota.Tipo", "<>", "C")
        .whereBetween("MestreNota.DtEmissao", [inicio, fim])
        .where({ "MestreNota.CdCliente": id[0] })
        .whereRaw(
          "ItemNota.CdFilial = MestreNota.CdFilial and ItemNota.Serie = MestreNota.Serie and ItemNota.Modelo = MestreNota.Modelo and ItemNota.NumNota = MestreNota.NumNota and CadPro.CdCategoria = CategPro.CdCategoria and NatOpe.Nop = MestreNota.Nop and CadVen.CdVendedor = MestreNota.CdVendedor and CadPro.CdProduto = ItemNota.CdProduto"
        )

        .groupBy("CdCliente", "NmCategoria")
        .groupByRaw("CONVERT(char(7), dbo.MestreNota.DtEmissao, 126)")

        .orderByRaw("CONVERT(CHAR(7),[mestrenota].[DtEmissao],126) DESC");

      return qry;
    },
  };
}
