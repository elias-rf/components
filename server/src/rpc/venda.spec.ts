import { Connections } from "dal/connections";
import Knex from "knex";
import { getTracker, MockClient, Tracker } from "knex-mock-client";
import { afterEach, beforeEach, describe, expect, test } from "vitest";
import { Venda } from "./venda";

describe("venda", () => {
  const knexDb = Knex({ client: MockClient });
  const rpc = Venda({
    plano: knexDb,
  } as Connections);
  let tracker: Tracker;

  beforeEach(() => {
    tracker = getTracker();
  });

  afterEach(() => {
    tracker.reset();
  });

  test("analitico", async () => {
    tracker.on.select("MestreNota").response(["ok"]);

    const rsp = await rpc.vendaAnalitico({
      inicio: "2020-01-01",
      fim: "2020-01-31",
    });

    expect(rsp).toEqual(["ok"]);
    expect(tracker.history.select[0].bindings).toEqual([
      2,
      "S",
      "S",
      "S",
      "N",
      0,
      3158,
      "E",
      "S",
      "2020-01-01",
      "2020-01-31",
    ]);
    console.log(
      `ðŸš€ ~ file: venda.spec.ts ~ line 44 ~ test ~ tracker.history.select[0].sql`,
      tracker.history.select[0].sql
    );
    expect(tracker.history.select[0].sql).toMatchSnapshot();
  });

  test("diario", async () => {
    tracker.on.select("MestreNota").response(["ok"]);

    const rsp = await rpc.vendaDiario({
      inicio: "2020-01-01",
      fim: "2020-01-31",
      uf: ["MG"],
    });

    expect(rsp).toEqual(["ok"]);
    expect(tracker.history.select[0].bindings).toEqual([
      "S",
      2,
      "S",
      "S",
      "N",
      0,
      "C",
      3158,
      "2020-01-01",
      "2020-01-31",
      "MG",
    ]);
    expect(tracker.history.select[0].sql).toMatchSnapshot();
  });

  test("mensal", async () => {
    tracker.on.select("MestreNota").response(["ok"]);

    const rsp = await rpc.vendaMensal({
      inicio: "2020-01-01",
      fim: "2020-01-31",
      id: ["10"],
    });

    expect(rsp).toEqual(["ok"]);
    expect(tracker.history.select[0].bindings).toEqual([
      "S",
      2,
      "S",
      "S",
      "N",
      0,
      "C",
      "2020-01-01",
      "2020-01-31",
      "10",
    ]);

    expect(tracker.history.select[0].sql).toEqual(
      `select "CategPro"."NmCategoria", "MestreNota"."CdCliente", CONVERT(char(7), MestreNota.DtEmissao, 126) AS anoMes, sum(case MestreNota.tipo when 'E' then ItemNota.Quantidade * -1 when 'S' then ItemNota.Quantidade end) as quantidade, sum(itemNota.VlLiquido * (case MestreNota.tipo when 'E' then ItemNota.Quantidade * -1 when 'S' then ItemNota.Quantidade end)) AS valor from MestreNota, ItemNota, CadVen, NatOpe, CadPro, CategPro where "MestreNota"."FgEstatistica" = ? and "MestreNota"."CdFilial" = ? and "CadPro"."FgEstatistica" = ? and "CadVen"."FgControle" = ? and "ItemNota"."ImprimeComponentes" = ? and "ItemNota"."Sequencia" > ? and "MestreNota"."Tipo" <> ? and "MestreNota"."DtEmissao" between ? and ? and "MestreNota"."CdCliente" = ? and ItemNota.CdFilial = MestreNota.CdFilial and ItemNota.Serie = MestreNota.Serie and ItemNota.Modelo = MestreNota.Modelo and ItemNota.NumNota = MestreNota.NumNota and CadPro.CdCategoria = CategPro.CdCategoria and NatOpe.Nop = MestreNota.Nop and CadVen.CdVendedor = MestreNota.CdVendedor and CadPro.CdProduto = ItemNota.CdProduto group by "CdCliente", "NmCategoria", CONVERT(char(7), dbo.MestreNota.DtEmissao, 126) order by CONVERT(CHAR(7),[mestrenota].[DtEmissao],126) DESC`
    );
  });
});
