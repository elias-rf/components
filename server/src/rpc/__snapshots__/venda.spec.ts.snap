// Vitest Snapshot v1

exports[`venda > analitico 1`] = `"select 'FV' as origem, \\"CategPro\\".\\"NmCategoria\\", \\"MestreNota\\".\\"DtEmissao\\", \\"MestreNota\\".\\"NumNota\\", \\"MestreNota\\".\\"Serie\\", \\"MestreNota\\".\\"Tipo\\", \\"ItemNota\\".\\"CdProduto\\", \\"ItemNota\\".\\"Quantidade\\", \\"ItemNota\\".\\"VlTotal\\", \\"CadPro\\".\\"Descricao\\", \\"MestreNota\\".\\"CdVendedor\\", \\"CadVen\\".\\"NmVendedor\\", \\"CidadesERF\\".\\"uf\\" from \\"MestreNota\\" inner join \\"ItemNota\\" on \\"ItemNota\\".\\"CdFilial\\" = \\"MestreNota\\".\\"CdFilial\\" and \\"ItemNota\\".\\"NumNota\\" = \\"MestreNota\\".\\"NumNota\\" and \\"ItemNota\\".\\"Serie\\" = \\"MestreNota\\".\\"Serie\\" and \\"ItemNota\\".\\"Modelo\\" = \\"MestreNota\\".\\"Modelo\\" inner join \\"CadVen\\" on \\"MestreNota\\".\\"CdVendedor\\" = \\"CadVen\\".\\"CdVendedor\\" inner join \\"CadCli\\" on \\"MestreNota\\".\\"CdCliente\\" = \\"CadCli\\".\\"CdCliente\\" inner join \\"cidadesErf\\" on \\"CadCli\\".\\"Cidade\\" = \\"cidadesErf\\".\\"NmCidadeIBGE\\" inner join \\"CadPro\\" on \\"ItemNota\\".\\"CdProduto\\" = \\"CadPro\\".\\"CdProduto\\" inner join \\"CategPro\\" on \\"CadPro\\".\\"CdCategoria\\" = \\"CategPro\\".\\"CdCategoria\\" where \\"MestreNota\\".\\"CdFilial\\" = ? and \\"CadVen\\".\\"FgControle\\" = ? and \\"MestreNota\\".\\"FgEstatistica\\" = ? and \\"CadPro\\".\\"FgEstatistica\\" = ? and \\"ItemNota\\".\\"ImprimeComponentes\\" = ? and \\"ItemNota\\".\\"Sequencia\\" <> ? and \\"mestrenota\\".\\"cdcliente\\" <> ? and \\"MestreNota\\".\\"DtEmissao\\" between ? and ? and \\"MestreNota\\".\\"Tipo\\" in (?, ?)"`;

exports[`venda > diario 1`] = `"select \\"NmCategoria\\", \\"MestreNota\\".\\"DtEmissao\\", sum(itemNota.Quantidade * (case MestreNota.tipo when 'E' then ItemNota.VlLiquido * -1 else ItemNota.VlLiquido end)) as valor, sum(\\"ItemNota\\".\\"Quantidade\\") as \\"quantidade\\" from MestreNota, ItemNota, CadVen, NatOpe, CadPro, CadCli, CategPro, CidadesERF where \\"MestreNota\\".\\"FgEstatistica\\" = ? and \\"MestreNota\\".\\"CdFilial\\" = ? and \\"CadPro\\".\\"FgEstatistica\\" = ? and \\"CadVen\\".\\"FgControle\\" = ? and \\"ItemNota\\".\\"ImprimeComponentes\\" = ? and \\"ItemNota\\".\\"Sequencia\\" > ? and \\"MestreNota\\".\\"Tipo\\" <> ? and \\"mestrenota\\".\\"cdcliente\\" <> ? and \\"MestreNota\\".\\"DtEmissao\\" between ? and ? and ItemNota.CdFilial = MestreNota.CdFilial and ItemNota.Serie = MestreNota.Serie and ItemNota.Modelo = MestreNota.Modelo and ItemNota.NumNota = MestreNota.NumNota and CadPro.CdCategoria = CategPro.CdCategoria and NatOpe.Nop = MestreNota.Nop and CadVen.CdVendedor = MestreNota.CdVendedor and CadCli.CdCliente = MestreNota.CdCliente and CadPro.CdProduto = ItemNota.CdProduto and Cadcli.Cidade = cidadeserf.NmCidadeIBGE and \\"cidadeserf\\".\\"uf\\" in (?) group by \\"NmCategoria\\", \\"MestreNota\\".\\"DtEmissao\\" order by \\"NmCategoria\\" asc, \\"MestreNota\\".\\"DtEmissao\\" desc"`;
